Sources/Swift/Integrations/SessionReplay/SentryTouchTracker.swift

```diff
@@ -0,0 +1,149 @@
+        for touch in touches {
+            guard touch.phase == .began || touch.phase == .ended || touch.phase == .moved || touch.phase == .cancelled else { continue }
+            let info = trackedTouches[touch] ?? TouchInfo(id: touchId++)
+            let position = touch.location(in: nil).applying(scale)
+            let newEvent = TouchEvent(x: position.x, y: position.y, timestamp: event.timestamp, phase: touch.phase.toRRWebTouchPhase())
+            
+            switch touch.phase {
+            case .began:
+                info.start = newEvent
+            case .ended, .cancelled:
+                info.end = newEvent
+            case .moved:
+                if let last = info.movements.last, touchesDelta(last.point, position) < 10 { continue }
+                info.movements.append(newEvent)
+                debounceEvents(in: info)
+            default:
+                continue
+            }
+            
+            trackedTouches[touch] = info
+        }
+    }
+    
+    private func touchesDelta(_ lastTouch: CGPoint, _ newTouch: CGPoint) -> CGFloat {
+        let dx = newTouch.x - lastTouch.x
+        let dy = newTouch.y - lastTouch.y
+        return sqrt(dx * dx + dy * dy)
+    }
+    
+    private func debounceEvents(in touchInfo: TouchInfo) {
+        guard touchInfo.movements.count >= 3 else { return }
+        let subset = touchInfo.movements.suffix(3)
+        if subset[subset.startIndex + 2].timestamp - subset[subset.startIndex + 1].timestamp > 0.5 {
+            //Dont debounce if the touch has a 1 second difference to show this pause in the replay.
+            return
+        }
+        if arePointsCollinearSameDirection(subset[subset.startIndex].point, subset[subset.startIndex + 1].point, subset[subset.startIndex + 2].point) {
+            touchInfo.movements.remove(at: touchInfo.movements.count - 2)
+        }
+    }
+    
+    private func arePointsCollinearSameDirection(_ a: CGPoint, _ b: CGPoint, _ c: CGPoint, tolerance: CGFloat = 10) -> Bool {
+        let crossProduct = (b.y - a.y) * (c.x - b.x) - (c.y - b.y) * (b.x - a.x)
+        
+        if abs(crossProduct) >= tolerance {
+            return false //Not collinear, return early
+        }
+              
+        let directionAB = CGPoint(x: b.x - a.x, y: b.y - a.y)
+        let directionBC = CGPoint(x: c.x - b.x, y: c.y - b.y)
+        let lengthAB = sqrt(pow(directionAB.x, 2) + pow(directionAB.y, 2))
+        let lengthBC = sqrt(pow(directionBC.x, 2) + pow(directionBC.y, 2))
+        let normalizedAB = CGPoint(x: directionAB.x / lengthAB, y: directionAB.y / lengthAB)
+        let normalizedBC = CGPoint(x: directionBC.x / lengthBC, y: directionBC.y / lengthBC)
+        let dotProduct = normalizedAB.x * normalizedBC.x + normalizedAB.y * normalizedBC.y
+        
+        // Directions are considered the same if the dot product is close to 1 (with some tolerance)
+        return abs(dotProduct - 1) < 0.001
```

Comment:
`m`: Out of curiosity? On which math or algorithm is this based?

Reply:
Some planar trigonometry. First half to detect if the lines are colinear (have the same angle), the second half is to detect if the lines go in the same direction.

Reply:
Now I get it. The code now is easier to understand. Again if you add some comments it's way easier to understand.